src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -163,5 +176,107 @@ private IEnumerable<CodeRefactoringProvider> GetProviders(Document document)
 
             return null;
         }
+
+        private ImmutableArray<CodeRefactoringProvider> GetProjectRefactorings(Project project)
+        {
+            // TODO (https://github.com/dotnet/roslyn/issues/4932): Don't restrict refactorings in Interactive
+            return project.Solution.Workspace.Kind == WorkspaceKind.Interactive
+                ? ImmutableArray<CodeRefactoringProvider>.Empty
+                : _projectRefactoringsMap.GetValue(project.AnalyzerReferences, pId => new StrongBox<ImmutableArray<CodeRefactoringProvider>>(ComputeProjectRefactorings(project))).Value;
+        }
+
+        private ImmutableArray<CodeRefactoringProvider> ComputeProjectRefactorings(Project project)
+        {
+            var extensionManager = project.Solution.Workspace.Services.GetService<IExtensionManager>();
+            ImmutableArray<CodeRefactoringProvider>.Builder builder = null;
+            foreach (var reference in project.AnalyzerReferences)
+            {
+                var projectCodeRefactoringProvider = _analyzerReferenceToRefactoringsMap.GetValue(reference, _createProjectCodeRefactoringsProvider);
+                foreach (var refactoring in projectCodeRefactoringProvider.GetRefactorings(project.Language))
+                {
+                    builder ??= ImmutableArray.CreateBuilder<CodeRefactoringProvider>();
+                    builder.Add(refactoring);
+                }
+            }
+
+            if (builder is null)
+            {
+                return ImmutableArray<CodeRefactoringProvider>.Empty;
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private class ProjectCodeRefactoringProvider
+        {
+            private readonly AnalyzerReference _reference;
+            private ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>> _refactoringsPerLanguage;
+
+            public ProjectCodeRefactoringProvider(AnalyzerReference reference)
+            {
+                _reference = reference;
+                _refactoringsPerLanguage = ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>>.Empty;
+            }
+
+            public ImmutableArray<CodeRefactoringProvider> GetRefactorings(string language)
+            {
+                return ImmutableInterlocked.GetOrAdd(ref _refactoringsPerLanguage, language, CreateRefactorings);
+            }
+
+            private ImmutableArray<CodeRefactoringProvider> CreateRefactorings(string language)
+            {
+                // check whether the analyzer reference knows how to return fixers directly.
+                if (_reference is ICodeRefactoringProviderFactory codeRefactoringProviderFactory)
```

Comment:
is there an example of when that would be the case?

Reply:
Copied the pattern from `CodeFixService` which uses this interface as a test hook.